components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt

```diff
@@ -283,3 +291,172 @@ private fun getMaxAvailableHeightToTopAndBottom(anchor: View): Pair<Int, Int> {
 
     return distanceToTop to distanceToBottom
 }
+
+/**
+ * Measure, calculate, obtain all data needed to know how the menu should be positioned.
+ *
+ * This method assumes [currentData] already contains the [MenuPositioningData.askedOrientation].
+ *
+ * @param containerView the menu layout that will be wrapped in the PopupWindow.
+ * @param anchor view the PopupWindow will be aligned to.
+ * @param currentData current known data for how the menu should be positioned.
+ *
+ * @return new [MenuPositioningData] containing the current constraints of the PopupWindow.
+ */
+@VisibleForTesting
+internal fun inferMenuPositioningData(
+    containerView: ViewGroup,
+    anchor: View,
+    currentData: MenuPositioningData
+): MenuPositioningData {
+    // Measure the menu allowing it to expand entirely.
+    val spec = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
+    containerView.measure(spec, spec)
+
+    val (availableHeightToTop, availableHeightToBottom) = getMaxAvailableHeightToTopAndBottom(anchor)
+    val containerHeight = containerView.measuredHeight
+
+    val fitsUp = availableHeightToTop >= containerHeight
+    val fitsDown = availableHeightToBottom >= containerHeight
+
+    return inferMenuPosition(anchor, currentData.copy(
+        fitsUp = fitsUp,
+        fitsDown = fitsDown,
+        availableHeightToTop = availableHeightToTop,
+        availableHeightToBottom = availableHeightToBottom,
+        containerViewHeight = containerHeight
+    ))
+}
+
+/**
+ * Infer where and how the PopupWindow should be shown based on the data available in [currentData].
+ * Should be called only once per menu to be shown.
+ *
+ * @param anchor view the PopupWindow will be aligned to.
+ * @param currentData current known data for how the menu should be positioned.
+ *
+ * @return new MenuPositioningData updated to contain the inferred [MenuPlacement]]
+ */
+@VisibleForTesting
+internal fun inferMenuPosition(anchor: View, currentData: MenuPositioningData): MenuPositioningData {
+    // Try to use the preferred orientation, if doesn't fit fallback to the best fit.
+
```

Comment:
Nit: blank line

Reply:
I actually intended that, to have the comment as the start of the method - a general explanation for future devs looking at the code and then actually laying the code.